os/fs/smartfs/smartfs_smart.c

```diff
@@ -712,22 +719,26 @@ static ssize_t smartfs_write(FAR struct file *filep, const char *buffer, size_t
 				goto errout_with_semaphore;
 			}
 
-			/* If file is modified and more data remains to be appended to file, but no next sector is available,
-			 * do not update sf->currsector and curroffset. Will be handled when buffer data is synced.
+			/* If file is modified and more data remains to be appended to file,
+			 * but no next sector is available, do not update sf->currsector and curroffset.
+			 * This will be handled when buffer data is synced.
 			 */
 			if (SMARTFS_NEXTSECTOR(header) != 0xFFFF) {
-				sf->currsector = SMARTFS_NEXTSECTOR(header);
-				
 				/* Now get the chained sector info and reset the offset */
+				sf->currsector = SMARTFS_NEXTSECTOR(header);
 				sf->curroffset = size;
+
 #ifdef CONFIG_SMARTFS_USE_SECTOR_BUFFER
-				/* Read the next sector if more bytes remain in the write buffer */
+				/* Read the next sector's data if we have reached the end of the current sector
+				 * AND more data is left to be written.
+				 * The header of the current sector will be read later if required.
+				 */
 				if (buflen > 0) {
-					smartfs_setbuffer(&readwrite, sf->currsector, 0,\
-						 fs->fs_llformat.availbytes, (uint8_t *)sf->buffer);
+					smartfs_setbuffer(&readwrite, sf->currsector, size,\
```

Comment:
you can improve here.
Let's think about buflen is greater than fs->fs_llformat.availbytes - size or smaller than it.